src/terminal/adapter/adaptDispatch.cpp

```diff
@@ -514,6 +514,9 @@ bool AdaptDispatch::CursorSaveState()
     savedCursorState.IsOriginModeRelative = _modes.test(Mode::Origin);
     savedCursorState.Attributes = attributes;
     savedCursorState.TermOutput = _termOutput;
+    savedCursorState.CursorType = textBuffer.GetCursor().GetType();
```

Comment:
We should refer to the standards if this is allowed:
https://vt100.net/docs/vt510-rm/DECSC.html
https://vt100.net/docs/vt510-rm/DECRC.html

Do you know what other terminals do in this case? Do they restore the cursor style too?

Reply:
ahh, didn't even realize there was a spec for this.  (I think we can close this if it doesn't match)   Do I have it right that it doesn't say if the cursor style should be saved or not?

Did a quick pass over some emulators.
conhost - restored the blinking _
alacrity - did not restore the blinking |
kitty - did restore the blinking |
wezterm - did restore the blinking |

Reply:
I think there are two things at play:
1. Switching Alternate buffer -> Main buffer.
2. Cursor save and restore.

And I suspect (1) is responsible for restoring the cursor style.

To test the theory:
1. Started with blinking block cursor:
```pwsh
pwsh $ "`e[1 q"
```
2. Then ran the command which:
   1. Activates the alternate buffer.
   2. Sets the cursor to steady block.
   4. Activates the main buffer again.

```pwsh
pwsh $ "`e[?1049h `e[2 q `e[?1049l"
```
3. I can still see the blinking cursor.

That means at least wezterm keeps a copy of the cursor style to restore when the alternate buffer exits.

---

And DECSC & DECRC doesn't restore cursor style even on Wezterm:

Again, if I start with a blinking cursor, then save the cursor, change the cursor to a steady block, and restore the cursor, the result is a steady block, not the old style.

```pwsh
pwsh $ "`e7 `e[2 q `e8"
```

I've only Wezterm installed so I couldn't test the other terminals 😅

Reply:
I think I understand now,  AdaptDispatch::CursorRestoreState is the implementation of DECSC/DECSR and is only meant to restore the cursor position and not the cursor style.  (Will see if I can figure out how wezterm and kitty are handling the cursor style and alt/main buffer)

It looked like Alacrity and Kitty both just moved the cursor when I ran this.
"`e7 `e[2 q `e8"

One thing that is weird is that my local build with the change didn't restore the blinking cursor when I ran.  (Will try to understand this also)
"`e7 `e[2 q `e8"

Reply:
> It looked like Alacrity and Kitty both just moved the cursor when I ran this.
```"`e7 `e[2 q `e8"```

That is to be expected as I added extra spaces to visually differentiate the escape codes in the comment 😅. You can use ```"`e7`e[2 q`e8"``` instead (note the single space here which is a required one).

Reply:
I think I kind of understand what is going on now.  I tried setting a breakpoint on AdaptDispatch::CursorSaveState() and tried pwsh $ "`e7" and noticed that the break point wasn't hit.

My theory is that powershell wasn't sending the DECSC.

I tried the following in kitty and observerd the position and style was restored this time using bash.
echo -e "\e[4 q"
echo -e "\e7"
echo -e "\e[2 q"
echo -e "\e8"

I tried the same with wezterm using bash instead of powershell and observed the same.

Looking at the source code, it looks like they do use DECSC/DECSR to save/restore the cursor style. (Will try to confirm the same for kitty)

https://github.com/wez/wezterm/blob/main/termwiz/src/cell.rs#L63
https://github.com/wez/wezterm/blob/main/termwiz/src/cell.rs#L45
https://github.com/wez/wezterm/blob/main/term/src/terminalstate/mod.rs#L136
https://github.com/wez/wezterm/blob/main/term/src/terminalstate/mod.rs#L2594
https://github.com/wez/wezterm/blob/main/term/src/terminalstate/mod.rs#L2637

I guess this doesn't answer the question of if DECSC/DECSR should be overload to save/resotre the style also.

Weztem with bash.
![wezterm-bash](https://github.com/microsoft/terminal/assets/811029/a3aaeb27-3c54-4d27-aaa3-494455f73134)